drain3/masking.py

```diff
@@ -1,35 +1,72 @@
 """
-Description : This file implements the persist/restore from Kafka
+Description : This file implements the masking using user-provided patterns
 Author      : Moshik Hershcovitch
 Author_email: moshikh@il.ibm.com
 License     : MIT
 """
-import logging
 import re
-from typing import List
+import abc
+from typing import Container
 
-logger = logging.getLogger(__name__)
 
+class AbstractMaskingInstruction(abc.ABC):
 
-class MaskingInstruction:
-    def __init__(self, regex_pattern: str, mask_with: str):
-        self.regex_pattern = regex_pattern
+    def __init__(self, mask_with: str):
         self.mask_with = mask_with
-        self.regex = re.compile(regex_pattern)
 
+    @abc.abstractmethod
+    def mask(self, content: str, mask_prefix: str, mask_suffix: str) -> str:
+        """
+        Mask content according to this instruction and return the result.
 
-class RegexMasker:
-    def __init__(self, masking_instructions: List[MaskingInstruction], mask_prefix: str, mask_suffix: str):
+        :param content: text to apply masking to
+        :param mask_prefix: the prefix of any masks inserted
+        :param mask_suffix: the suffix of any masks inserted
+        """
+        pass
+
+
+class MaskingInstruction(AbstractMaskingInstruction):
+
+    def __init__(self, pattern: str, mask_with: str):
+        super().__init__(mask_with)
+        self.regex = re.compile(pattern)
+
+    @property
+    def pattern(self):
+        return self.regex.pattern
+
+    def mask(self, content: str, mask_prefix: str, mask_suffix: str) -> str:
+        mask = mask_prefix + self.mask_with + mask_suffix
+        return self.regex.sub(lambda x: mask, content)
```

Comment:
no need for lambda here

Reply:
According to https://docs.python.org/3/library/re.html#re.sub having unknown escapes in the replacement-value with sub is no longer supported. 
If either `mask_prefix`, `mask_with` or `mask_suffix` contains a backslash (or backslash + ASCII) this will result in errors.
If a function is used as a callback this is not a problem.
As this was not possible before the refactoring I can also just remove this; alternatively I can add a test to guarantee this behavior.

Reply:
i dont think we should support name names with `\` in them